src/features/wallet/helpers/collectRewards.ts

```diff
@@ -32,46 +33,80 @@ export default function collectRewards(
+            ? datum.ancillaryData
+            : "0x";
+          pendingRequest.identifier = ethers.utils.toUtf8Bytes(
+            datum.identifier
+          );
+          pendingRequest.time = datum.time;
+          pendingRequestData.push(pendingRequest);
+        }
+      });
+
+      postData.pendingRequests = pendingRequestData;
+      mcr.target = contract.address;
+      mcr.callData = votingInterface.encodeFunctionData(
+        "retrieveRewards(address,uint256,(bytes32,uint256,bytes)[])",
+        [postData.voterAddress, postData.roundId, postData.pendingRequests]
+      );
 
-    data.forEach((datum) => {
-      if (datum.roundId === roundId) {
-        const pendingRequest = {} as PendingRequestRetrieveReward;
-        postData.roundId = datum.roundId;
-        postData.voterAddress = datum.address;
-        pendingRequest.ancillaryData = datum.ancillaryData
-          ? datum.ancillaryData
-          : "0x";
-        pendingRequest.identifier = ethers.utils.toUtf8Bytes(datum.identifier);
-        pendingRequest.time = datum.time;
-        pendingRequestData.push(pendingRequest);
-      }
+      multicallCollectRewards.push(mcr);
     });
 
-    postData.pendingRequests = pendingRequestData;
-    mcr.target = contract.address;
-    mcr.callData = votingInterface.encodeFunctionData(
-      "retrieveRewards(address,uint256,(bytes32,uint256,bytes)[])",
-      [postData.voterAddress, postData.roundId, postData.pendingRequests]
+    return (
+      multicallContract.functions["aggregate((address,bytes)[])"](
+        multicallCollectRewards
+      )
+        // wait for at least 1 block conf.
+        .then((tx) => tx.wait(1))
+        .then((conf: any) => {
+          setAvailableRewards(DEFAULT_BALANCE);
+        })
+        .catch((err) => console.log("err in retrieve rewards", err))
     );
+  } else {
+    const postData = {} as PostRetrieveReward;
+    const pendingRequestData = [] as PendingRequestRetrieveReward[];
 
-    multicallCollectRewards.push(mcr);
-  });
+    uniqueRoundIds.forEach((roundId) => {
+      data.forEach((datum) => {
+        if (datum.roundId === roundId) {
+          const pendingRequest = {} as PendingRequestRetrieveReward;
+          postData.roundId = datum.roundId;
+          postData.voterAddress = datum.address;
+          pendingRequest.ancillaryData = datum.ancillaryData
+            ? datum.ancillaryData
+            : "0x";
+          pendingRequest.identifier = ethers.utils.toUtf8Bytes(
+            datum.identifier
+          );
+          pendingRequest.time = datum.time;
+          pendingRequestData.push(pendingRequest);
+        }
+      });
 
-  // console.log("multi calls sorted?", multicallCollectRewards);
+      postData.pendingRequests = pendingRequestData;
+    });
 
-  return (
-    multicallContract.functions["aggregate((address,bytes)[])"](
-      multicallCollectRewards
-    )
-      // wait for at least 1 block conf.
+    return retrieveRewards(contract, postData)
```

Comment:
would be nice if we could functionalize these further so `collectRewards` could call `collectManyRewards` and `collectOneReward` which both return a promise where you can de-duplicate promise handling logic:
```js
let collect // typing this might be difficult
if (uniqueRoundIds.length > 1) {
  collect = collectManyRewards(data, uniqueRoundIds,pendingRequests, ...etc)
}else{
  collect = collectOneRewards(data, uniqueRoundIds,pendingRequests, ...etc)
}
collect.then((tx) => tx.wait(1))
      .then((conf: any) => {
        setAvailableRewards(DEFAULT_BALANCE);
      })
      .catch((err) => console.log("err in retrieve rewards", err));
```

Reply:
Wouldn't it be simpler to return those functions?

Reply:
Maybe we should move the set visual state into the wallet file anyway.